### ISSUE TITLE
No hot reloading or connection to JS runtime with :npm-module build target

### ISSUE THREAD

**Issue by bpringe** [thheller/shadow-cljs#1094]
Would it be difficult to add support for connecting the repl to the JS runtime with :npm-module as well as adding hot-reload?

We discussed some of this in [this Slack thread](https://clojurians.slack.com/archives/C6N245JGG/p1676500815612249).

---

Here are some notes of an attempt made to get a repl connection, with your guidance, in a test project, on [this branch](https://github.com/BetterThanTomorrow/calva-cljs-testbed/tree/attempt_1):

I updated the shadow-cljs.edn file to look like:

```edn
{:deps true
 :builds       {:extension
                {:target :npm-module
                 :runtime :node
                 :output-dir "node_modules/shadow-cljs"
                 :entries [calva-cljs.extension
                           calva.foo]
                 :devtools {:before-load-async calva-cljs.extension/before-load-async
                            :after-load calva-cljs.extension/after-load}}}}
```

I added an `index.js` file for the extension entry point that looks like:

```js
console.log("Loading shadow devtools for node client");
require("shadow-cljs/shadow.cljs.devtools.client.node");
console.log("Loading extension entry point");
const extension = require("shadow-cljs/calva_cljs.extension");

exports.activate = extension.activate;
exports.deactivate = extension.deactivate;
```

I set the `main` property in `package.json` to `"./index.js"`.

When I run the extension and run the `Hello World` command, it works. However, when I save the `extension.cljs` file, I get the following error in the debug console (though, the shadow-cljs compilation succeeds):

```text
JS reload failed ReferenceError: SHADOW_IMPORT is not defined
```

You suggested copying the contents of the `node_bootstrap.js` file from the shadow-cljs source into the entry js file. After doing that, the above-mentioned procedure leads to an error that `SHADOW_IMPORT_PATH` is not defined. It looks like the `shadow.build.node/flush-unoptimized` adds a definition for that variable, but I'm guessing this function isn't called for an :npm-module build. When I tried to set it to `__dirname + "/node_modules/shadow-cljs"`, which seemed correct for our setup, another error occurred:

```text
SHADOW import error /Users/brandon/development/calvacljstestbed/node_modules/shadow-cljs/calva_cljs.extension.js
extensionHostProcess.js:100
JS reload failed Error: Cannot find module './cljs_env'
Require stack:
- /Users/brandon/development/calvacljstestbed/index.js
...
```

That file/module is defined at the relative path mentioned, so it's not clear what to do next.

---

Might there be a simple way to get what I'm trying to do working without changing shadow-cljs's source? If not, do you know if adding support for this in shadow-cljs would be difficult / much work?

CC @pez

**Comment by thheller** [comment]
Where is the branch with the `:npm-module` attempt?

**Comment by thheller** [comment]
See https://github.com/thheller/calva-cljs-testbed/tree/attempt_1

I can compile everything and run `F5` in vscode. Running the Command seems to work, and the REPL appears to work, at least from Cursive.

**Comment by bpringe** [comment]
Thanks!

Responses to some of your comments from the readme:

> Most of the troubles of previous attempts came from trying to mix two builds, or in that case even three because of the TS.

Yes, I eventually abandoned trying to use two builds after having so much trouble and talking with you about it.

> `import from "goog:..."` is the way the closure compiler expects JS files to load Closure namespaces, which CLJS namespaces also are. I do not know how you tell tsc to ignore these.

I found that we can get rid of those tsc errors by adding this comment above those types of imports: `// @ts-ignore`. Another way to get rid of the errors is to add a file (named something like `types.d.ts`) and declare the modules like so:

```typescript
declare module "goog:calva.foo";
```

One issue we have with Calva is that because of how the `vscode` import works (I don't know the details off hand), none of the unit tested code can import `vscode`, so currently the cljs used in Calva does not import `vscode`, and if we do import `vscode` in the cljs then we get an error when trying to run tests that `vscode` can't be found (or something like that).

This is because some of the TS code being tested imports the cljs lib file like `import * as cljsLib from '<some-path/cljs-lib.js'`. This is why I was trying to use the `:npm-module` target, to separate compiled cljs code that imports `vscode` from that which doesn't. 

I'm wondering if using the `goog:*` import method with the `:node-library` target will only pull in the code for that namespace even though all of it still exists in a single file, allowing us to still run unit tests but import `vscode` in other, new cljs namespaces that won't be pulled in by the TS being unit tested. I'll have to test this out.

**Comment by thheller** [comment]
Can you give code examples for the problems you are describing? I'm sure they can be fixed, but without seeing code in the context it is used in, it is very hard to follow rough descriptions. Important details are likely missing.

Unit tests are an entirely different subject and will most likely not use the `:node-library` build. It is ok to have multiple builds for development purposes, as long as they are not loaded into the same runtime simultanously.

> I'm wondering if using the goog:* import method with the :node-library target will only pull in the code for that namespace even though all of it still exists in a single file

I'm unsure what you mean here. The `goog:` trick is like importing any other namespace, it will bring it that namespace and all of its transitive dependencies. It will not bring in namespaces that aren't referenced.

One thing I'm unclear on: Right now calva is primarily driven by TS code, CLJS code is imported as a library. Do you just intend to increase the amount of CLJS code doing stuff? Or is your intent to make CLJS the primary driver (aka. extension entry point) and slowly replace the TS code? The setup I created makes CLJS the primary driver and that has certain implications on how you do stuff. You can go either way, how you do stuff just changes.

**Comment by bpringe** [comment]
> Can you give code examples for the problems you are describing?

Regarding the unit testing issue with importing `vscode`, you can replicate the issue on [this Calva branch](https://github.com/BetterThanTomorrow/calva/tree/add_new_output_window) by doing this:

1. `npm i`
2. `npm run clean`
3. `npm run compile-cljs`
4. `npm run unit-test`

You'll get the following error:

```text
SHADOW import error /Users/brandon/development/calva/.shadow-cljs/builds/calva-lib/dev/out/cljs-runtime/shadow.js.shim.module$vscode.js

Error: Cannot find module 'vscode'
Require stack:
- /Users/brandon/development/calva/out/cljs-lib/cljs-lib.js
- /Users/brandon/development/calva/src/calva-fmt/src/format-index.ts
```

This is because `src/extension-test/unit/formatting-test.ts` imports `src/calva-fmt/src/format-index.ts`, which imports the  compiled cljs-lib js file, which is importing `vscode` because `src/cljs-lib/src/calva/repl_output.cljs` requires `vscode`, and `vscode` cannot imported in unit-tested code because it's provided by the VS Code extension host environment. See [this issue](https://github.com/microsoft/vscode-test/issues/37) for more details on that, if interested.

**In that issue I see a potential workaround for us though - mocking `vscode`, which might be worth trying.** CC @PEZ

---

>> I'm wondering if using the goog:* import method with the :node-library target will only pull in the code for that namespace even though all of it still exists in a single file

> I'm unsure what you mean here. The goog: trick is like importing any other namespace, it will bring it that namespace and all of its transitive dependencies. It will not bring in namespaces that aren't referenced.

I think you answered my question here. What I meant relates to the error mentioned above. The `:node-library` build creates one file, and if we import using a relative file path (at least) as is happening in the situation above, and any of the cljs code in that build imports `vscode`, we get an error like the one above. 

If the `goog:*` import really does _only_ bring in the code for that namespace, even though all of the compiled cljs code is in one js file, then maybe that could get rid of the error for us. However, we may have issues running our hundreds on TS unit tests _as they are_ due to the `goog:*` imports. I don't yet know if we can work around that, aside from maybe rewriting the tests in cljs, which would be a large endeavor.

---

> One thing I'm unclear on: Right now calva is primarily driven by TS code, CLJS code is imported as a library. Do you just intend to increase the amount of CLJS code doing stuff? Or is your intent to make CLJS the primary driver (aka. extension entry point) and slowly replace the TS code?

My intention is to increase the amount of CLJS code doing stuff, which includes being able to use `vscode` from cljs, and to be able to do so incrementally, until ideally one day all/most of Calva is CLJS. Whichever method makes it easiest for us to achieve that over time is fine with me. The two options I see are:

1. Keep TS at the top level / keep tsc as the build tool, and grow the CLJS from the bottom up until most of Calva is CLJS. This is only really a good option if we can use `vscode` in the CLJS and still run our hundreds of unit tests, since it would mean we can add new features basically wholly in CLJS, and just activate and/or call into the CLJS feature code from TS where necessary.
2. Make shadow-cljs the primary driver, basically sandwiching the TS in between CLJS, and then move TS code to CLJS code until the TS is gone. The benefit here is that we get hot reload our existing TS code because shadow is the driver. I guess ideally we use this solution.

**Comment by thheller** [comment]
OK, this highlights why I asked the question about the primary driver.

In your current test setup TS is the primary driver, it drives the tests via mocha, which seems to be calling the TS compiler? I assume you don't want to rewrite all your tests, so `:node-library` is basically out. I think as a migration strategy `:npm-module` is then the better fit, as you can basically move code one namespace/file at a time. 

However, in this kind of a setup 2 way interop becomes the problem. In the `:node-library` setup shadow-cljs is responsible for building everything. It can see all TS code and all CLJS code, it can stitch everything together to make one coherent system. With `:npm-module` it can only see the CLJS code. It has no knowledge about any of the npm/TS code. So, while the TS code can easily import the CLJS code, the CLJS code cannot directly reference the TS code. There are of course ways to still access the TS code, but not directly via `ns` `:require`.

I'll see about an example setup for that.

**Comment by bpringe** [comment]
> In your current test setup TS is the primary driver, it drives the tests via mocha, which seems to be calling the TS compiler?

Correct, as far as I understand. This is why I was referring to rewriting tests, which is not something we want to do, as you assumed.

> I think as a migration strategy :npm-module is then the better fit, as you can basically move code one namespace/file at a time.

Yeah, this is why I started gravitating toward this, but when I try this I can't get a repl connection or hot reload, and without those two things I don't care to go this route. But, of course, this goes back to the original focus of this issue. If we can get hot reload and a repl connection, I think this is a method worth using.

> So, while the TS code can easily import the CLJS code, the CLJS code cannot directly reference the TS code.

Yeah I think this is acceptable, and it's how I assumed it would be.

> I'll see about an example setup for that.

If you can set up an example that includes hot reload of the cljs and allows a repl connection, that would be awesome and much appreciated.

**Comment by thheller** [comment]
I spent some time with this yesterday. The node REPL is indeed entirely broken for `:npm-module`.

The current node REPL implementation is designed specificly for `:node-script` and `:node-library` targets, which combine all output into one file. The `__filename`, `__dirname`, `require` and `module` specials that node commonjs modules have access too are handled in a way that they all act like the final output would. Otherwise `__filename` would yield different results in dev and release builds, which has caused issues in the past. `require` behavior is even worse. So, I'm not going to change that implementation, especially given that `__filename` or rather `js/__filename` must yield the same result regardless of loading it via the runtime, hot-reload or REPL.

However for `:npm-module` this assumption does not hold. Each namespace sits in its own file, for development and in release builds. So, each should have their own `__filename`, `require`, `module`. `__dirname` is technically always the same, given there is only one output-dir. So, an entire new implementation seems warranted. I'll see if I can implement one that works universally, so `:runtime` is no longer a required parameter at all.

**Comment by bpringe** [comment]
Thanks!

I ran into a couple of issues when trying out your repo at the currently latest commit (c7cd87d2010c1f4783ed95e0b8c2816a756e4956).

I run the project like so:

1. `npm run watch-ts` (I get an error about the cljs-lib not being found at that path, but I assume that is to be expected)
2. I jack-in, which runs `npx shadow-cljs -d cider/cider-nrepl:0.28.5 watch :ext` (I updated shadow-cljs in the package.json, too, in case that was necessary for the changes to be used by this command. I noticed it was updated in deps.edn already.)
3. Start the extension host

When I try to run the "Hello cljs-lib" command I get an error: `Cannot read properties of undefined (reading 'test_function')`.

It seems that `cljsLib` is undefined [here](https://github.com/thheller/calva-cljs-testbed/blob/c7cd87d2010c1f4783ed95e0b8c2816a756e4956/src/ts/bar.ts#L14).

When I load extension.cljs, I get an error: `Cannot find module '../bar.js'`. However, when I evaluate `(.hello bar)`, I get `"Hello from TypeScript"`.

**Comment by bpringe** [comment]
> When I try to run the "Hello cljs-lib" command I get an error: Cannot read properties of undefined (reading 'test_function').

> It seems that cljsLib is undefined [here](https://github.com/thheller/calva-cljs-testbed/blob/c7cd87d2010c1f4783ed95e0b8c2816a756e4956/src/ts/bar.ts#L14).

I have fixed the above issue by changing the import statement from `import cljsLib ...` to `import * as cljsLib ...`.

> When I load extension.cljs, I get an error: Cannot find module '../bar.js'

This is still an issue, but I think we can could just not use the compiled ts from cljs.

In your repo I also get an error from TS compiler like this:

```text
src/ts/bar.ts:3:26 - error TS2307: Cannot find module './cljs/calva.foo' or its corresponding type declarations.

3 import * as cljsLib from "./cljs/calva.foo";
```

But I know this is because the path is relative to the output directory.

If we continue to keep the TS at the top level / as the extension entrypoint, then I think we can just send the shadow build output to node_modules and import cljs like `... from 'shadow-cljs/calva.foo` in the TS.

**Comment by bpringe** [comment]
We can fix the `Cannot find module './cljs/calva.foo'` error by using `require` instead of `import`, like:

```typescript
const cljsLib = require("./cljs/calva.foo");
```

**Comment by thheller** [comment]
What a gigantic pile of **** all this TS stuff is. For reasons I do not understand it only seems to output the `lib/bar.js` file like 50% of the time. Sometimes it is written, sometimes it is not. Don't ask me why, but yeah things will not work if `lib/bar.js` is missing.

Whatever you do you MUST NOT move the output to `node_modules`. I mean not if you want to use TS code from CLJS, which I thought was a base requirement. If not then yes, you can make things a whole lot easier by just moving it to `node_modules`.

**Comment by thheller** [comment]
Ah, it seems to be related to the `"incremental":true` option in `tsconfig.json`. As soon as I remove that it seems to output the files properly. Still with the warnings, but at least it actually writes the files.

**Comment by bpringe** [comment]
I still get the error `Cannot find module '../bar.js'` when I load extension.cljs, but the calls to bar.js work.

I noticed I do not get hot reloading of the cljs. Is that to be expected? If I change what the "Hello World" command shows in the message, save the file, see the cljs compilation succeed, then run the "Hello World" command again, I don't get the updated message.

**Comment by bpringe** [comment]
> Whatever you do you MUST NOT move the output to node_modules. I mean not if you want to use TS code from CLJS

So, I know you said this 😄, but I did find a solution that sort of works like this. Though, I don't know if some issues might occur with packaging or something, given that the `main` property in package.json is set to a file in node_modules.

The solution is in [this branch](https://github.com/BetterThanTomorrow/calva-cljs-testbed/tree/attempt_1) as of commit `46ed078`. I get a repl connection and hot reload of the cljs and the ts.

My shadow-cljs.edn:

```edn
{:deps true
 :builds       {:extension
                {:target :npm-module
                 :runtime :node
                 :output-dir "node_modules/shadow-cljs"
                 :entries [calva-cljs.extension
                           calva.foo]
                 :devtools {:before-load-async calva-cljs.extension/before-load-async
                            :after-load calva-cljs.extension/after-load}}}}
```

My TS output goes to `out`, and I added `out` to the paths in deps.edn. In the `calva-cljs.extension` I required `/foo` in the ns form.

The only issue I've seen is that I can't evaluate stuff from `foo.js` in the repl in `calva-cljs.extension`. The command that call the `hello` function from `foo` works, but if I evaluate `(.. foo (hello))` in the repl, I get an exception: `module$foo.hello is not a function`.

Anyway, I'm thinking maybe just going forward with the cljs not calling the compiled ts is fine. It definitely seems to be a simpler path in terms of maintenance over time and the dev experience for us and contributors.

**Comment by thheller** [comment]
> I still get the error Cannot find module '../bar.js' when I load extension.cljs, but the calls to bar.js work.

Please upload the branch that has this. I do not understand what you mean. There should be a `out/bar.js`. The CLJS output should be in `out/cljs`. Therefore any CLJS file can access `out` via the `../` prefix. That is all there is. It also doesn't get any easier that this. I intentially changed your TS example file from `foo` to `bar` since it got increasingly confusing to have a calva.foo and a foo.

I though the whole point of this exercise was keeping TS files out of the CLJS build, so that tests and stuff still work? In your setup shadow-cljs processes the TS code. I all I can say is that I **very strongly** suggest not doing your setup, it will not work for the whole calva codebase. shadow-cljs or rather `:advanced` will mess the TS code too much, and you'll spend days trying to debug externs issues. Unless, of course you are fine with `:simple`, at which point the problem goes away.

Keeping the TS code out of the CLJS build avoids a whole category of problems. Letting it resolve at runtime is just safer. As soon as you put `out` in the classpath you are basically asking for trouble. Also by putting code into `node_modules/shadow-cljs` you are making the code "unbundlable", so since you I guess you are planning to publish this somewhere this is not an option.

**Comment by bpringe** [comment]
> Please upload the branch that has this.

This is with your repo. The only thing I changed was the import of `calva.foo` in bar.ts to `import * as cljsLib from "./cljs/calva.foo"` to avoid the issue of `cljsLib` being undefined. It's just an error when loading the cljs file, but the require is actually working, so it's not a huge deal.

Here's a video. I load the file, you see the error, then I evaluate the require directly, you see the error again, then I evaluate calls to `bar.js` and they work.

https://user-images.githubusercontent.com/12722744/224513993-3f793778-0069-4acb-8868-d8e512c5c3a5.mov


This isn't really something we need to solve though as I don't think we'll go the route of requiring js in the cljs.

---

> I though the whole point of this exercise was keeping TS files out of the CLJS build, so that tests and stuff still work?

I think the test problem was only a problem with the :npm-module target when we used the `goog:*` requires of the cljs code in the ts. If we're using relative paths then the ts compiler can find the code. I haven't fully verified that last point yet but I'm pretty confident that's true.

> I all I can say is that I very strongly suggest not doing your setup, it will not work for the whole calva codebase. shadow-cljs or rather :advanced will mess the TS code too much, and you'll spend days trying to debug externs issues. Unless, of course you are fine with :simple, at which point the problem goes away.

I assume you're talking about the setup with shadow building the whole extension, including the ts / compiled ts code. I think :simple would be okay if we went this route, but I don't think we will. Going that route seems to add unnecessary complexity in terms of maintenance, without that much benefit over the route of just keeping the ts at the top level, out of the cljs code, and building the extension as we do now.

> Keeping the TS code out of the CLJS build avoids a whole category of problems.

Thanks for the guidance. I think we will keep the ts out of the cljs.

> Also by putting code into node_modules/shadow-cljs you are making the code "unbundlable", so since you I guess you are planning to publish this somewhere this is not an option.

Do you mean this with any approach, not just if shadow is building the extension?

**Comment by thheller** [comment]
Since I don't have actual Calva installed, and wouldn't know how to use it I went through a REPL session in Cursive. Evaling the forms in the comment block, then load-file. And repeating it just for good measure. All works fine. No issues.

```
Connecting to remote nREPL server...
Clojure 1.11.1
(shadow/repl :ext)
To quit, type: :cljs/quit
=> [:selected :ext]
=> "Hello from TypeScript"
=> "Hello from cljs-lib"
=> "Hello from cljs-lib"
Loading src/main/calva/extension.cljs... done
=> "Hello from TypeScript"
=> "Hello from cljs-lib"
=> "Hello from cljs-lib"
=>
[#js{:hello #object[hello],
     :platform #object[platform],
     :cljsLibTestFunction #object[cljsLibTestFunction],
     :showMessage #object[showMessage]}]
Loading src/main/calva/extension.cljs... done
```

`Loading src/main/calva/extension.cljs... done` meaning that a `load-file` for the file was issued.

`[#js{ ... }]` being the evaluated `(js/require "../bar.js")`, which is the datafied var. Just noticed that it is doing that when it shouldn't.

Not sure what could be different? I only did the same change you did.

`node_modules/shadow-cljs` doesn't work since it is the `shadow-cljs` npm package as far as node/npm is concerned. You cannot put your files into the `shadow-cljs` package that I publish. So, whenever you publish your `calva` extension, it will be missing the necessary files in the `node_modules/shadow-cljs` package.

Although thinking about that now the calva extension probably isn't published to npm is it? You just create a zip file with all the files? So, in theory that might work, but shadow-cljs still becomes part the published `calva` package, which is unnecesary. Just put it into its own folder its not something you need to think about.

**Comment by bpringe** [comment]
Hmm, yeah I'm not sure what might be different, but at least the code works as expected aside from the error that just makes it look like there's an issue.

> Although thinking about that now the calva extension probably isn't published to npm is it? You just create a zip file with all the files?

Correct, it's not published on npm. And yes, it's essentially packaged with all the needed files. I don't know for sure what would happen when packaging if the compiled cljs was in `node_modules/shadow-cljs`, but we can just keep them outside of `node_modules`.

Thanks so much for your help with all this.
